Code/Framework/AzCore/AzCore/EBus/EBusSharedDispatchMutex.h

```diff
@@ -0,0 +1,187 @@
+     *   - Bus connects / disconnects will only execute when no event dispatches are executing
+     *   - Event dispatches can call other event dispatches on the same bus recursively
+     *
+     * Limitations:
+     *   - If the bus contains custom connect / disconnect logic, it must not call any event dispatches on the same bus.
+     *   - Bus connects / disconnects cannot happen within event dispatches on the same bus.
+     *
+     * Usage:
+     *   Add code like the following to the definition of a specific EBus:
+     *      using MutexType = AZ::EBusSharedDispatchMutex;
+     *
+     *      template <typename MutexType, bool IsLocklessDispatch>
+     *      using DispatchLockGuard = AZ::EBusSharedDispatchMutexDispatchLockGuard<AZ::EBus<MyBus>>;
+     *
+     *      template<typename MutexType>
+     *      using ConnectLockGuard = AZ::EBusSharedDispatchMutexConnectLockGuard<AZ::EBus<MyBus>>;
+     *
+     *      template<typename MutexType>
+     *      using CallstackTrackerLockGuard = AZ::EBusSharedDispatchMutexCallstackLockGuard<AZ::EBus<MyBus>>;
+     */
+
+    // Simple custom mutex class that contains a shared_mutex for use with connects / disconnects / event dispatches
+    // and a separate mutex for callstack tracking thread protection.
+    class EBusSharedDispatchMutex
+    {
+    public:
+        EBusSharedDispatchMutex() = default;
+        ~EBusSharedDispatchMutex() = default;
+
+        void CallstackMutexLock()
+        {
+            m_callstackMutex.lock();
+        }
+
+        void CallstackMutexUnlock()
+        {
+            m_callstackMutex.unlock();
+        }
+
+        void EventMutexLockExclusive()
+        {
+            m_eventMutex.lock();
+        }
+
+        void EventMutexUnlockExclusive()
+        {
+            m_eventMutex.unlock();
+        }
+
+        void EventMutexLockShared()
+        {
+            m_eventMutex.lock_shared();
+        }
+
+        void EventMutexUnlockShared()
+        {
+            m_eventMutex.unlock_shared();
+        }
+
+    private:
+        AZStd::shared_mutex m_eventMutex;
+        AZStd::mutex m_callstackMutex;
+
+        // This custom mutex type should only be used with the lock guards below since it needs additional context to know which
+        // mutex to lock and what type of lock to request. If you get a compile error due to these methods being private, the EBus
+        // declaration is likely missing one or more of the lock guards below.
+        void lock(){}
+        void unlock(){}
+    };
+
+    // Custom lock guard to handle Connection lock management.
+    // This will lock/unlock the event mutex with an exclusive lock. It will assert and disallow
+    // exclusive locks if currently inside of a shared lock.
+    template<class EBusType>
+    class EBusSharedDispatchMutexConnectLockGuard
+    {
+    public:
+        AZ_FORCE_INLINE EBusSharedDispatchMutexConnectLockGuard(EBusSharedDispatchMutex& mutex, AZStd::adopt_lock_t)
```

Comment:
Should we avoid adding new instances of AZ_FORCE_INLINE?

Reply:
I put them here because they were used on lock_guard, scoped_lock, and unique_lock, and these classes are just custom variations of those. Happy to remove them though if anyone feels they should be. They shouldn't make _any_ difference in this case since having the implementation in the class definition should already force inlining.

Reply:
Removed.